File: node/deliberation/src/observers/hard-block-cascade.mjs
/**
 * @accretive-substrate/deliberation/observers/hard-block-cascade
 *
 * Detects gates that hard-block deliberations repeatedly. Pattern: same
 * gate name appearing in guardrail_result.hard_blocks N+ times in 30min.
 * Emits a `gate-def:<gate>` accretion suggesting threshold review.
 */

import { getPool } from '../store.mjs';

const DEFAULT_N_THRESHOLD = 3;
const DEFAULT_WINDOW_MINUTES = 30;
const COOLDOWN_MINUTES = 30;

let _accretiveModule = null;
async function _loadAccretive() {
  if (_accretiveModule) return _accretiveModule;
  try { _accretiveModule = await import('@accretive-substrate/accretive'); } catch { _accretiveModule = null; }
  return _accretiveModule;
}

export async function detectHardBlockCascade(options = {}) {
  const pool = options.pool || getPool();
  if (!pool) return [];
  const N = options.nThreshold ?? DEFAULT_N_THRESHOLD;
  const win = options.windowMinutes ?? DEFAULT_WINDOW_MINUTES;
  try {
    const r = await pool.query(`
      SELECT
        block_name,
        COUNT(*)::int AS n,
        ARRAY_AGG(d.id ORDER BY d.ts) AS deliberation_ids,
        MIN(d.ts) AS first_seen_at,
        MAX(d.ts) AS last_seen_at
      FROM deliberations d, jsonb_array_elements_text(d.guardrail_result->'hard_blocks') AS block_name
      WHERE d.ts > NOW() - ($1 || ' minutes')::interval
        AND d.final_verdict = 'abort'
      GROUP BY block_name
      HAVING COUNT(*) >= $2`,
      [String(win), N],
    );
    return r.rows || [];
  } catch (e) {
    console.error(`[observer:hard-block-cascade] query failed: ${e.message}`);
    return [];
  }
}

async function _isInCooldown(pool, canonicalPath, cooldownMinutes) {
  try {
    const r = await pool.query(
      `SELECT 1 FROM accretions WHERE canonical_path = $1
        AND provenance_class = 'coach_provisional'
        AND appended_at > NOW() - ($2 || ' minutes')::interval LIMIT 1`,
      [canonicalPath, String(cooldownMinutes)],
    );
    return r.rows.length > 0;
  } catch { return false; }
}

export async function runHardBlockCascadeObserver(options = {}) {
  const pool = options.pool || getPool();
  if (!pool) return { detected: 0, emitted: 0, skipped: 0, errors: 0 };
  let appendAccretion = options.appendAccretion;
  if (!appendAccretion) {
    const mod = await _loadAccretive();
    if (!mod) return { detected: 0, emitted: 0, skipped: 0, errors: 1, reason: '@accretive-substrate/accretive unavailable' };
    appendAccretion = mod.appendAccretion;
  }

  const detections = await detectHardBlockCascade(options);
  let emitted = 0, skipped = 0, errors = 0;
  const cooldown = options.cooldownMinutes ?? COOLDOWN_MINUTES;
  const win = options.windowMinutes ?? DEFAULT_WINDOW_MINUTES;

  for (const d of detections) {
    const path = `gate-def:${d.block_name}`;
    if (await _isInCooldown(pool, path, cooldown)) { skipped++; continue; }
    const fact = `Gate ${d.block_name} hard-blocked ${d.n} times in last ${win}min — coach proposes threshold/override review`;
    const result = await appendAccretion({
      canonical_path: path,
      operator: 'coach',
      fact,
      provenance_class: 'coach_provisional',
      operator_confirmed: false,
      raw: {
        evidence_refs: {
          deliberation_ids: d.deliberation_ids.map(Number),
          n_occurrences: d.n,
          window: `${win}min`,
          first_seen_at: d.first_seen_at,
          last_seen_at: d.last_seen_at,
          gate_id: d.block_name,
        },
        observer: 'hard-block-cascade',
        observer_version: '1.0.0',
      },
    });
    if (result.ok) emitted++; else errors++;
  }
  return { detected: detections.length, emitted, skipped, errors };
}

File: go/deliberation/pkg/deliberation/observers/hardblockcascade/hardblockcascade.go
// Package hardblockcascade — Go observer for gates that hard-block N+ times.
package hardblockcascade

import (
	"context"
	"fmt"

	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/ramene/accretive-substrate/go/deliberation/pkg/deliberation/observers"
)

const Name = "hard-block-cascade"

type Detection struct {
	BlockName       string  `json:"block_name"`
	N               int     `json:"n"`
	DeliberationIDs []int64 `json:"deliberation_ids"`
	FirstSeenAt     string  `json:"first_seen_at"`
	LastSeenAt      string  `json:"last_seen_at"`
}

func Detect(ctx context.Context, pool *pgxpool.Pool, n, windowMinutes int) ([]Detection, error) {
	if n <= 0 {
		n = observers.DefaultNThreshold
	}
	if windowMinutes <= 0 {
		windowMinutes = observers.DefaultWindowMinutes
	}
	rows, err := pool.Query(ctx, `
		SELECT block_name, COUNT(*)::int,
		       ARRAY_AGG(d.id ORDER BY d.ts), MIN(d.ts)::text, MAX(d.ts)::text
		FROM deliberations d, jsonb_array_elements_text(d.guardrail_result->'hard_blocks') AS block_name
		WHERE d.ts > NOW() - (($1)::int || ' minutes')::interval
		  AND d.final_verdict = 'abort'
		GROUP BY block_name
		HAVING COUNT(*) >= $2`, windowMinutes, n,
	)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := []Detection{}
	for rows.Next() {
		var d Detection
		if err := rows.Scan(&d.BlockName, &d.N, &d.DeliberationIDs, &d.FirstSeenAt, &d.LastSeenAt); err != nil {
			return nil, err
		}
		out = append(out, d)
	}
	return out, nil
}

func Run(ctx context.Context, pool *pgxpool.Pool) observers.Result {
	out := observers.Result{Observer: Name}
	if pool == nil {
		return out
	}
	detected, err := Detect(ctx, pool, observers.DefaultNThreshold, observers.DefaultWindowMinutes)
	if err != nil {
		out.Errors = 1
		return out
	}
	out.Detected = len(detected)
	for _, d := range detected {
		path := "gate-def:" + d.BlockName
		if observers.IsInCooldown(ctx, pool, path, observers.DefaultCooldownMinutes) {
			out.Skipped++
			continue
		}
		fact := fmt.Sprintf("Gate %s hard-blocked %d times in last %dmin — coach proposes threshold/override review",
			d.BlockName, d.N, observers.DefaultWindowMinutes)
		evidence := map[string]any{
			"deliberation_ids": d.DeliberationIDs,
			"n_occurrences":    d.N,
			"window":           fmt.Sprintf("%dmin", observers.DefaultWindowMinutes),
			"first_seen_at":    d.FirstSeenAt,
			"last_seen_at":     d.LastSeenAt,
			"gate_id":          d.BlockName,
		}
		if err := observers.EmitProvisional(ctx, pool, path, fact, Name, evidence); err != nil {
			out.Errors++
			continue
		}
		out.Emitted++
	}
	return out
}
